### PR TITLE
Introduce fisheye plugin

### PR DESCRIPTION
**Pull request by soreau**
This is a fisheye effect ported from compiz magnifier plugin fisheye fragment program.

### CODE REVIEW

File: plugins/single_plugins/fisheye.cpp [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+        t1 = t1 * p1 + p2;
+
+        tc = texture2D(u_texture, t1.xy).rgb;
+
+        gl_FragColor = vec4(tc, 1.0);
+}
+)";
+
+class wayfire_invert_screen : public wayfire_plugin_t
```
Comment (ammen99) [inline plugins/single_plugins/fisheye.cpp]: s/invert_screen/fisheye/ or something among these lines

File: plugins/single_plugins/fisheye.cpp [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+            GL_CALL(glBindTexture(GL_TEXTURE_2D, 0));
+
+            if (!duration.running() && target_zoom == 0)
+            {
+                output->render->rem_post(&hook);
+                hook_set = false;
+            }
+
+            if (active)
+                duration.start(current_zoom, target_zoom);
```
Comment (ammen99) [inline plugins/single_plugins/fisheye.cpp]: I can't get why you start at every frame ..

File: plugins/single_plugins/fisheye.cpp [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+}
+)";
+
+class wayfire_invert_screen : public wayfire_plugin_t
+{
+
+    post_hook_t hook;
+    key_callback toggle_cb;
+    wf_duration duration;
+    float radius, target_zoom, current_zoom;
```
Comment (ammen99) [inline plugins/single_plugins/fisheye.cpp]: why do you need current zoom? I think you can just use `duration.progress()` every time you need it, or make it a local variable for `render()`. Same for radius, just use `radius_opt->as_cached_double()` each time you need it, and in this case, you can rename it to just `radius`

File: plugins/single_plugins/fisheye.cpp [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+            GL_CALL(glDisable(GL_BLEND));
+            GL_CALL(glBindFramebuffer(GL_DRAW_FRAMEBUFFER, target));
+            GL_CALL(glDrawArrays (GL_TRIANGLE_FAN, 0, 4));
+            GL_CALL(glBindFramebuffer(GL_FRAMEBUFFER, 0));
+
+            GL_CALL(glDisableVertexAttribArray(posID));
+            GL_CALL(glBindTexture(GL_TEXTURE_2D, 0));
+
+            if (!duration.running() && target_zoom == 0)
```
[the remark this answers is not in the document]
Comment (soreau) [reply plugins/single_plugins/fisheye.cpp]: It was being set in toggle_cb() if active is true here https://github.com/soreau/wayfire/blob/ed66f7d87826c437a7545b1ca91b229c92389f00/plugins/single_plugins/fisheye.cpp#L110

File: plugins/single_plugins/fisheye.cpp (hunk 3 above)
Comment (soreau) [reply plugins/single_plugins/fisheye.cpp]: You're basically right, I just made current_zoom local to render() and got rid of radius_opt.

File: plugins/single_plugins/fisheye.cpp (hunk 2 above)
Comment (soreau) [reply plugins/single_plugins/fisheye.cpp]: This is so changing target_zoom via wcm works.

File: plugins/single_plugins/fisheye.cpp (hunk 1 above)
Comment (soreau) [reply plugins/single_plugins/fisheye.cpp]: Ack.

File: plugins/single_plugins/fisheye.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (soreau) [reply plugins/single_plugins/fisheye.cpp]: Actually that's current_zoom so you're right, it's never set to 0. I removed the condition.

**Comment by ammen99** [comment]
LGTM (code- and aesthetics-wise), thanks!

**Pull request merged**